src/main/java/com/example/bookstore/service/impl/ShoppingCartServiceImpl.java

```diff
@@ -0,0 +1,83 @@
+package com.example.bookstore.service.impl;
+
+import com.example.bookstore.dto.cartitem.CartItemRequestDto;
+import com.example.bookstore.dto.shoppingcart.ShoppingCartDto;
+import com.example.bookstore.mapper.CartItemMapper;
+import com.example.bookstore.mapper.ShoppingCartMapper;
+import com.example.bookstore.model.CartItem;
+import com.example.bookstore.model.ShoppingCart;
+import com.example.bookstore.model.User;
+import com.example.bookstore.repository.BookRepository;
+import com.example.bookstore.repository.CartItemRepository;
+import com.example.bookstore.repository.ShoppingCartRepository;
+import com.example.bookstore.repository.UserRepository;
+import com.example.bookstore.service.ShoppingCartService;
+import jakarta.persistence.EntityNotFoundException;
+import java.util.Optional;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class ShoppingCartServiceImpl implements ShoppingCartService {
+    private final BookRepository bookRepository;
+    private final UserRepository userRepository;
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final CartItemRepository cartItemRepository;
+    private final ShoppingCartMapper shoppingCartMapper;
+    private final CartItemMapper cartItemMapper;
+
+    @Override
+    public ShoppingCartDto getShoppingCartDto(String email) {
+        ShoppingCart shoppingCart = getShoppingCart(email);
+        shoppingCart.setCartItems(cartItemRepository.getAllByShoppingCart(shoppingCart));
+        return shoppingCartMapper.toDto(shoppingCart);
+    }
+
+    @Override
+    public ShoppingCartDto addItemToCart(String email, CartItemRequestDto requestDto) {
+        if (!bookRepository.existsById(requestDto.getBookId())) {
+            throw new EntityNotFoundException("Can't find book with id: "
+                    + requestDto.getBookId());
+        }
+        ShoppingCart shoppingCart = getShoppingCart(email);
+        CartItem cartItem = cartItemMapper.toModel(requestDto);
+        cartItem.setShoppingCart(shoppingCart);
+        cartItem.setBook(bookRepository.getReferenceById(requestDto.getBookId()));
```

Comment:
You see, now we are requesting the same book twice, what we can do is fetch it right away with `Optional<Book> findById(Long id)`, check that optional, and then assign it to the `cartItem`, thus avoiding 2 calls to the DB
By the way, a handy way to check it would be using the optional `orElseThrow()` method, making it more in the functional style :wink: